### PR TITLE
Update: PanelColorGradientSettings to use dropdowns

### PR DESCRIPTION
**Pull request by jorgefilipecosta**
Alternative to: https://github.com/WordPress/gutenberg/pull/37053

This PR updates PanelColorGradientSettings to be dropdown-based. The advantage when compared to https://github.com/WordPress/gutenberg/pull/37053 is that all the blocks even the ones that use a custom color implementation (navigation, cover, social links, and even third-party) automatically get the new UI.

## How has this been tested?
I verified the new color UI appears and works as expected on the paragraph, group, cover, navigation, and social links blocks.
I verified the global styles colors still work as before.

## Screenshots 
image
image
image
image

### CODE REVIEW

**Comment by shaunandrews** [comment]
Very nice. This feels a lot more natural than the drill-down option explored earlier. Its also easier to use when editing a few colors back-to-back; There's no need to find a back button, or wait for transitions.

--

I noticed the Solid/Gradient toggle seems broken when first opening the color menu:
image

--

I think we should likely look at a few more tweaks:

- Give the menu some more padding; It feels a little cramped right now.
- The currently selected color in the sidebar should have a different state to help reinforce which color you're changing. Maybe keep the text blue (like hover) and add a grey-100 background color.
- The menu might benefit from a title using the currently selected color, like "Background color."

image

File: packages/block-editor/src/components/colors-gradients/style.scss
Comment (jasmussen) [inline packages/block-editor/src/components/colors-gradients/style.scss]: 37028 has been merged, so these rules can be removed. That gets us to this point at least:
Screenshot 2021-12-02 at 16 59 51

The diagonal line is missing. I'm realizing that this diagonal is scoped to work only in global styles: https://github.com/WordPress/gutenberg/pull/36994/files#diff-448ace8cb198438b19f4fc13f3107589730cf0aea599c46fa935e87ff1b70816R54

Where's a good place to move that rule? The diagonal should be shown when a color is _unset_, and the transparent checkerboard can be shown when a color is explicitly transparent. Any thoughts?

**Comment by pablohoneyhoney** [comment]
You are all too fast and had me jumping from one PR to the other 😄 

Good notes from @shaunandrews above, some of which I list out below as well.  

- I believe the colors opened in pop over is fine as an interim path. It does at least relieve the footprint occupied today in favor of more contextual color editing/selection. 
This pop over could however benefit from some finessing in paddings and various details however (hex code position, swatches are larger than the ones in inspector, borders compared to the color picker, clear button, etc.). The notice/tip inherited styles are out of place. It seems the transparency was addressed in a separate PR but not applicable here as @jasmussen points out. Some refinement to do also in the gradients case (segmentation seems glitchy until I engage with it).

- The overlap of the color picker is a bit troubling. It’s a bit convoluted visually but also functionally. Due to its position (could it be positioned below or the left side of the color pop over if triggered from this instance? Below can be an issue in large color options), and because I can’t move it (there’s no dragging affordance). It’s also frustrating when I click the canvas and everything resets to collapsed, provided I’m in color editing mode. Color picker only goes away when I select the color pop over.

Screen Shot 2021-12-02 at 10 50 04 AM

I'm personally also ok compromising some of the above in the name of 5.9, and iterate further later, but there's likely some challenging a11y aspects switching between these panels. 

One temporary alt solution I don't love is to add a header to the pop over to close and drag. 
 
LMK if the above resonates with you or if anything further context is needed.

File: packages/block-editor/src/components/colors-gradients/style.scss
Comment (mtias) [reply packages/block-editor/src/components/colors-gradients/style.scss]: The vertical alignment of the circles also seems off in the above capture

**Comment by shaunandrews** [comment]
> One temporary alt solution I don't love is to add a header to the pop over to close and drag.

I briefly thought about this, but then considered we wouldn't be using a "menu" anymore. I think that might have more drastic implications to usability and accessibility. I'm not arguing against it, just that it would likely mean a different implementation, maybe?

File: packages/block-editor/src/components/colors-gradients/style.scss
Comment (jasmussen) [reply packages/block-editor/src/components/colors-gradients/style.scss]: A fix for that was merged in https://github.com/WordPress/gutenberg/pull/37028, I wonder if a rebase would fix it?

**Review by mirka — COMMENTED** [review]
I’m assuming we’re mostly talking about things in [SC 2.4.8 Location](https://www.w3.org/TR/UNDERSTANDING-WCAG20/navigation-mechanisms-location.html) in regards to the a11y implications of these double popovers. (As opposed to the things we already support, e.g. focus management and keyboard operability.)

I agree it feels pretty reasonable when the popover content is minimal, but in other cases we may be pushing it. For example in this mobile viewport with Gradient options, we are basically looking at three layers of popover with an obstructed tab panel in between:

Block inspector for the Cover block in a mobile viewport, with nested color popovers

That’s a pretty deep stack to keep in your head, though it could be somewhat alleviated by things like #35292, or positioning the popovers in a way that makes the origins clearer. At least in the mobile use case though, it could be argued that the navigator approach in #36952 (or anything that reduces the popover stack) makes more sense. (I can't actually try it because the branch is already deleted.)

Maybe relevant is something like the Figma color flyout, which is designed in a way that doesn't require a secondary popover.

https://user-images.githubusercontent.com/555336/144488374-c51b2198-9e1f-4f88-a043-fca50e6907d6.mp4

File: packages/block-editor/src/components/colors-gradients/panel-color-gradient-settings.js
Comment (jasmussen) [inline packages/block-editor/src/components/colors-gradients/panel-color-gradient-settings.js:161]: It looks like the addition of the Dropdown wrapper here adds an extra element around each item inside the itemgroup. So without the dropdown you have this markup:

Screenshot 2021-12-03 at 11 10 35

With the extra wrapper, we have this:
Screenshot 2021-12-03 at 11 10 55

In terms of tabbing that doesn't seem to be a problem due to the tabindex properties. But the extra wrapper does mean this rule now is causing trouble:

Screenshot 2021-12-03 at 11 12 09

The rule is around the use of `border` for focus inside the itemgroup, making it transparent if it's the last child in an itemgroup. But because of the extra container, it's now adding transparent borders to every dropdown container which makes them all 1px taller. Compare this context:
Screenshot 2021-12-03 at 11 09 32

... with that of global styles:
Screenshot 2021-12-03 at 11 09 26

My instinct tells me the solution is to refactor how focus works inside the itemgroup, to not rely on those advanced selectors, but perhaps instead just use a pseudo element. That would also let us avoid the use of transparent borders, which I think are problematic anyway due to them being made opaque in high contrast modes. 

But before I explore a fix for this in a separate PR, I wonder: @ciampo do you think my instinct above makes sense, or would there be a different or better way to fix this?

**Comment by jasmussen** [comment]
I took the liberty of rebasing this branch, and pushing a small fix to add the diagonal indicator for unset colors:
Screenshot 2021-12-03 at 11 28 26

Per [the comment here](https://github.com/WordPress/gutenberg/pull/37067/files#r761807702), I'll look separately at how/if we can fix the spacing issue separately in the itemgroup.

File: packages/block-editor/src/components/colors-gradients/panel-color-gradient-settings.js
Comment (jasmussen) [reply packages/block-editor/src/components/colors-gradients/panel-color-gradient-settings.js:161]: @jorgefilipecosta it looks like the `ItemGroup` component is designed to have a very specific structure of `ItemGroup > Item` (like `ul > li`) and that the border styles are designed for the DOM structure to be pretty specific. As noted above the addition of the `Dropdown` is causing a few small issues since the structure becomes `ItemGroup > Dropdown > Item`, making the border rules target the wrong element.

It's not problematic enough to block this PR, I think, but it's also not ideal.

Just to see if it was possible, I tried to see if it would work to move the `Dropdown` inside the `Item`. Like so:

```
			<ItemGroup isBordered isSeparated>
				{ settings.map( ( setting, index ) => (
					<Item
						key={ index }
						className="block-editor-panel-color-gradient-settings__item"
					>
						<Dropdown
							contentClassName="block-editor-panel-color-gradient-settings__dropdown"
							renderToggle={ ( { onToggle } ) => {
								return (
									<HStack
										justify="flex-start"
										onClick={ onToggle }
									>
										<FlexItem>
											<ColorIndicator
												colorValue={
													setting.gradientValue ??
													setting.colorValue
												}
											/>
										</FlexItem>
										<FlexItem>{ setting.label }</FlexItem>
									</HStack>
								);
							} }
							renderContent={ () => (
								<ColorGradientControl
									showTitle={ false }
									{ ...{
										colors,
										gradients,
										disableCustomColors,
										disableCustomGradients,
										__experimentalHasMultipleOrigins,
										__experimentalIsRenderedInSidebar,
										enableAlpha,
										...setting,
									} }
								/>
							) }
						/>
					</Item>
				) ) }
			</ItemGroup>
```

It doesn't totally work, it replaces the `UnstyledButton` with the `Dropdown`. But is there a way we could make it work? Any thoughts?

File: packages/block-editor/src/components/colors-gradients/style.scss
Comment (ciampo) [inline packages/block-editor/src/components/colors-gradients/style.scss]: We should not rely on internal components classnames, as they are not part of the public APIs and could change in the future.

A better approach would be to set a custom className in the specific context of the block editor's color panel, and then use that new custom classname to apply styles (both for dropdown and navigator screen)

File: packages/block-editor/src/components/colors-gradients/style.scss
Comment (ciampo) [inline packages/block-editor/src/components/colors-gradients/style.scss]: As @jasmussen mentioned [here](https://github.com/WordPress/gutenberg/pull/37053#discussion_r761188224), we should be able to remove this rule safely

File: packages/block-editor/src/components/colors-gradients/style.scss
Comment (ciampo) [reply packages/block-editor/src/components/colors-gradients/style.scss]: _(Ignore this comment in case we're deleting this rule)_

We should try to avoid using internal classnames for styling `@wordpress/components` (same as comment above)

File: packages/block-editor/src/components/colors-gradients/panel-color-gradient-settings.js
Comment (ciampo) [reply packages/block-editor/src/components/colors-gradients/panel-color-gradient-settings.js:161]: @jasmussen analysis is correct.

I've tried playing around with the markup a little, but I couldn't find any good solution — mainly trying to change the outline to something like:

```
ItemGroup
  Item
    Dropdown
      Button
        [content]
```

But that would mean that we'd need to apply most of the styles of `<Item>` to `<Button>` to achieve the same look&feel (and even then there could be some glitches)

I think that the best solution for now is to keep things as they are, and look into improving how `ItemGroup` and `Item` are styled directly at the source level.

**Review by ciampo — COMMENTED** [review]
> I agree it feels pretty reasonable when the popover content is minimal, but in other cases we may be pushing it. For example in this mobile viewport with Gradient options, we are basically looking at three layers of popover with an obstructed tab panel in between:

I tend to agree with the points made by @mirka here as well

**Comment by jorgefilipecosta** [comment]
Thank you all for the feedback it seems the technical changes are ok. I'm working on the design taking into account the feedback provided. 

For planning purposes, Is this change supposed to be part of WP 5.9 or it is something for the future?

**Comment by jasmussen** [comment]
I think what's in this branch is a vast improvement over what's shipping in trunk, and I'd personally love to see it.

**Comment by letasync** [comment]
This is a welcome UI enhancement.
Recently I was also playing with color palette UI and had this in mind. 
The problem I see with current suggestion is, its hiding the main content area. Anything which overlaps the main content area of editor will again create bad user experience.




color palette UI suggestion 01

**Comment by jorgefilipecosta** [comment]
Thank you all for the feedback! I tried to incorporate the design feedback.
This is what we have:

Untitled 13


image

I faced some challenges because of underlying issues on the ToggleGroupControl. I added "temporary solutions" in this PR so we can test things properly. All the changes related to this component will be extracted into separate PR's so we can discuss the changes in a more detailed and focused way. Please ignore changes related to this component from this PR.

Another issue is the custom gradient picker color popover not having a great sidebar behavior we will get that for "free" after https://github.com/WordPress/gutenberg/pull/37115 is merged so also please ignore that until the other PR is merged.

**Comment by jasmussen** [comment]
Took this for a spin, and high level this feels like a very big enhancement over what's shipping, where it's hard to figure out which color you're changing simply due to the multiple duplicated color palettes stacked.

colors

I think there's going to be room to improve this further. For example the "Clear" button remains a bit of an odd one, but this can be enhanced by the color panel becoming a tools panel.

I don't think we should remove the border around the popover, though. That full-bleed look works for the innermost color picker, and it might work for making the top swatch have that edge to edge look. But I think we should explore that separately, and keep the gray border in the mean time. That means [this mockup](https://user-images.githubusercontent.com/1204802/132485764-cf3ac49c-0c71-4167-ad57-ef518744ab15.png) from #35093 is probably the closest thing to target at this point.

**Comment by jasmussen** [comment]
I'm seeing some positioning issues with individual gradient control points:
Screenshot 2021-12-10 at 12 17 04

**Review by pumpi851 — COMMENTED** [review]
[8990V7-99VO]

File: packages/edit-site/src/components/global-styles/style.scss
Comment (pumpi851) [inline packages/edit-site/src/components/global-styles/style.scss:68]: update/panel-color-gradient-settings-to-use-dropdowns

**Review by pumpi851 — APPROVED** [review]
[8990V7-99VO]update/panel-color-gradient-settings-to-use-dropdowns

File: packages/edit-site/src/components/global-styles/style.scss
Comment (pumpi851) [reply packages/edit-site/src/components/global-styles/style.scss:68]: update/panel-color-gradient-settings-to-use-dropdowns
[Uploading Magisk-23.0.zip…]()

File: packages/edit-site/src/components/global-styles/style.scss
Comment (pumpi851) [inline packages/edit-site/src/components/global-styles/style.scss:74]: update/panel-color-gradient-settings-to-use-dropdowns

**Review by pumpi851 — APPROVED** [review]
update/panel-color-gradient-settings-to-use-dropdowns

**Review by pumpi851 — COMMENTED** [review]
[update/panel-color-gradient-settings-to-use-dropdowns](Google)

**Comment by jorgefilipecosta** [comment]
Hi @jasmussen,

> I'm seeing some positioning issues with individual gradient control points:

That issue got fixed because https://github.com/WordPress/gutenberg/pull/37115 was merged.


> I don't think we should remove the border around the popover, though. That full-bleed look works for the innermost color picker, and it might work for making the top swatch have that edge to edge look. But I think we should explore that separately, and keep the gray border in the mean time. That means this mockup from #35093 is probably the closest thing to target at this point.

I reverted the border removal.
We have side popover to pick a custom color, while on the mockup we have a navigation for custom. That would work well on block inspectior but we would need to disable on global styles so the same component would need multiple flags. Do you think the side popover could work as a temporary solution?

**Comment by jasmussen** [comment]
> Do you think the side popover could work as a temporary solution?

I think what we have in this branch already is better than what's shipping, and a good step forward.

**Review by jasmussen — APPROVED** [review]
After the latest changes, things are looking good with the behavior, borders, general appearance and tabbing flow:
tabbing through the dialog

While there are some small things here and there (as already noted on the thread), this is such an important step forward, I think we should land it.

**Comment by ciampo** [comment]
> I think we should land it.

Just noting that the comments that I left above haven't been addressed yet:
- https://github.com/WordPress/gutenberg/pull/37067#discussion_r762115907
- https://github.com/WordPress/gutenberg/pull/37067#discussion_r762116308
- https://github.com/WordPress/gutenberg/pull/37067#discussion_r762119130

I also keep being skeptical about the double popover approach — using @mirka's words:

> I agree it feels pretty reasonable when the popover content is minimal, but in other cases we may be pushing it. For example in this mobile viewport with Gradient options, we are basically looking at three layers of popover with an obstructed tab panel in between:

**Comment by jasmussen** [comment]
I think of this as a temporary improvement due to how bad the situation in trunk is with the multiple duplicated palettes and unclear subheading hierarchy. I would want us to follow up on the popover feedback which is highly valid, and hopefully something like a sliding interface as tracked in #35093 could help there.

**Comment by ciampo** [comment]
> I think of this as a temporary improvement due to how bad the situation in trunk is with the multiple duplicated palettes and unclear subheading hierarchy. I would want us to follow up on the popover feedback which is highly valid, and hopefully something like a sliding interface as tracked in #35093 could help there.

I see your point regarding the UX of this approach, and understand that this is not a definitive solution.

I would still prefer if the comments that I left regarding styles and hardcoded classnames were addressed before merging, though (they shouldn't take long)

**Comment by jasmussen** [comment]
>  would still prefer if the comments that I left regarding styles and hardcoded classnames were addressed before merging, though (they shouldn't take long)

Definitely. I should've clarified that I gave a green check based on the user experience improvement, and it still needs to pass the code review.

File: packages/block-editor/src/components/colors-gradients/style.scss
Comment (jorgefilipecosta) [reply packages/block-editor/src/components/colors-gradients/style.scss]: We need this rule because of the empty line but the class enhancement was done.

**Comment by jorgefilipecosta** [comment]
Hi @jasmussen, @ciampo, @mirka  thank you for the reviews.


> I also keep being skeptical about the double popover approach — using @mirka's words:

On mobile, it is not ideal, on the next iteration, it is probably something we should work on.

> I would still prefer if the comments that I left regarding styles and hardcoded classnames were addressed before merging, though (they shouldn't take long)

The suggestions regarding classes were applied.

**Comment by jorgefilipecosta** [comment]
I had to skip two tests end to end tests because I'm having some trouble fixing them. The functionality works well it is the tests that need an update to account for the UI changes. I'm working on fixing the tests and should have a PR ready soon.

**Comment by priethor** [comment]
Beta 3 is being released as I write this message, and Beta 4 is not yet confirmed, so we should be mindful of the timeline and concerns about adding improvements so close to RC. As much as I’d personally like to have this improvement in 5.9, could we compromise by removing the `Backport to WP` label and releasing it in a minor version instead, unless feedback coming from Beta 3 suggests otherwise?

**Comment by mtias** [comment]
I marked this one as backport to a minor release so we can follow up quickly. If during the remaining testing period of 5.9 this surfaces as an important pain point for people we can reconsider it since it's ready.

**Comment by Mamaduka** [comment]
Cherry-picked for 5.9.1.

**Comment by Humanify-nl** [comment]
I hope this is the right place for this: 

As a heavy user of the block editor, editing a lot of blocks per hour, I have one big issue with this change to the color tool:

It added **two extra clicks** to clear the color styles from blocks. One for text, one for background. This adds a lot of unnecessary interactions (and tiresome) when editing many blocks daily.


**Solution?** 
Add a 'X or clear' icon next to the color previewer (not in the pop-over, but in the inspector panel) so we can quickly remove these colors. This used to be 2 clicks, now its 4 (and extra mouse movement).

**Pull request merged**
